### PR TITLE
Add PIN and PUK management views

### PR DESCRIPTION
**Pull request by emlun**
This PR depends on Yubico/yubikey-manager#157.

### CODE REVIEW

File: ykman-gui/qml/PivPinManagementView.qml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        visible: running
+    }
+
+    CustomContentColumn {
+        visible: !isBusy
+
+        ColumnLayout {
+            Layout.alignment: Qt.AlignLeft | Qt.AlignTop
+            Heading1 {
+                text: qsTr("Configure PIV PINs")
```
Comment (dagheyman) [inline ykman-gui/qml/PivPinManagementView.qml]: Let's use "Configure PINs" as a heading, thats what the button and breadcrumb say.

File: ykman-gui/qml/ChangePinView.qml [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                TextField {
+                    id: currentPin
+                    Layout.fillWidth: true
+                    echoMode: TextInput.Password
+                    selectByMouse: true
+                    selectionColor: yubicoGreen
+                    visible: hasCurrentPin
+                }
+
+                CustomButton {
```
Comment (dagheyman) [inline ykman-gui/qml/ChangePinView.qml]: I don't think we should have a default buttons here, users with uninitialised devices should use the init wizard

File: ykman-gui/qml/ChangePinView.qml [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    readonly property bool pinMatches: newPin.text === confirmPin.text
+
+    function validPin() {
+        return (pinMatches) && (chosenPin.length >= minLength)
+                && (chosenPin.length <= maxLength)
+    }
+
+    signal changePin(string currentPin, string newPin)
+    signal clearPinInputs
+    signal goBack
```
[the remark this answers is not in the document]
Comment (emlun) [reply ykman-gui/qml/ChangePinView.qml]: I felt it best to decouple this component from the specifics of the navigation hierarchy since it's a generic component.

File: ykman-gui/qml/Constants.qml [hunk 4]
```diff
@@ -3,6 +3,10 @@ import QtQuick 2.9
 QtObject {
     readonly property int fido2PinMinLength: 4
     readonly property int fido2PinMaxLength: 128
+    readonly property int pivPinMinLength: 4
```
[the remark this answers is not in the document]
Comment (emlun) [reply ykman-gui/qml/Constants.qml]: oops

File: ykman-gui/qml/ChangePinView.qml (hunk 3 above)
[the remark this answers is not in the document]
Comment (emlun) [reply ykman-gui/qml/ChangePinView.qml]: Agreed offline that this is unnecessary generality.

File: ykman-gui/qml/PivPinManagementView.qml (hunk 1 above)
Comment (emlun) [reply ykman-gui/qml/PivPinManagementView.qml]: Ok!

File: ykman-gui/qml/ChangePinView.qml (hunk 2 above)
Comment (emlun) [reply ykman-gui/qml/ChangePinView.qml]: Agreed offline to keep the default buttons until we figure out how to solve the initialization flow in the face of touch-requiring management keys.

Comment (dagheyman) [reply ykman-gui/qml/ChangePinView.qml]: But also, let's try with checkboxes instead, since the value is always the same. Like `Use Serial` in the Yubico OTP view.

File: ykman-gui/qml/ChangePinView.qml (hunk 3 above)
[the remark this answers is not in the document]
Comment (emlun) [reply ykman-gui/qml/ChangePinView.qml]: Fixed!

File: ykman-gui/qml/Constants.qml (hunk 4 above)
[the remark this answers is not in the document]
Comment (emlun) [reply ykman-gui/qml/Constants.qml]: Fixed!

File: ykman-gui/qml/PivPinManagementView.qml [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        yubiKey.refreshPiv(function () {
+            isBusy = false
+        })
+    }
+
+    function getPinMessage() {
+        if (pinBlocked) {
+            return qsTr("PIN is blocked.")
+        } else {
+            return qsTr("%1 PIN retries left").arg(pinRetries)
```
Comment (dagheyman) [inline ykman-gui/qml/PivPinManagementView.qml]: Lets go with "retries left" like in the fido view

File: ykman-gui/qml/PivPinManagementView.qml (hunk 1 above)
Comment (emlun) [reply ykman-gui/qml/PivPinManagementView.qml]: Fixed!

File: ykman-gui/qml/PivPinManagementView.qml (hunk 5 above)
Comment (emlun) [reply ykman-gui/qml/PivPinManagementView.qml]: Fixed!

File: ykman-gui/qml/ChangePinView.qml (hunk 2 above)
Comment (emlun) [reply ykman-gui/qml/ChangePinView.qml]: Fixed!

File: ykman-gui/py/yubikey.py [hunk 6]
```diff
@@ -132,6 +147,17 @@ def write_config(self, usb_applications, nfc_applications, lock_code):
             logger.error('Failed to write config', exc_info=e)
             return {'success': False, 'error': str(e)}
 
+    def refresh_piv(self):
+        with self._open_piv() as piv_controller:
+            return {
+                'has_protected_key': piv_controller.has_protected_key,
+                'pin_tries': piv_controller.get_pin_tries(),
+                'supported_touch_policies': [
```
Comment (dagheyman) [inline ykman-gui/py/yubikey.py]: Not sure we want to implement pin and touch policies to begin with, they are not enabled by default in PIV manager.

File: ykman-gui/qml/YubiKey.qml [hunk 7]
```diff
@@ -26,6 +26,8 @@ Python {
     property bool yubikeyModuleLoaded: false
     property bool yubikeyReady: false
     property var queue: []
+    property var piv
```
Comment (dagheyman) [inline ykman-gui/qml/YubiKey.qml]: These should probably live in the PivView.qml for now, or we should move out all data variables from the views.

File: ykman-gui/py/yubikey.py [hunk 8]
```diff
@@ -34,6 +34,17 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+class PivContextManager:
```
Comment (dagheyman) [inline ykman-gui/py/yubikey.py]: Since we decided offline that we want this, I'll try to prepare a PR that does something similar for OTP and FIDO.

File: ykman-gui/py/yubikey.py (hunk 6 above)
Comment (emlun) [reply ykman-gui/py/yubikey.py]: Good point, I'll remove the parts from here that aren't currently used.

**Comment by dagheyman** [comment]
FYI, cherry-picked over the context manager stuff to the `piv` branch because it seemed unrelated and I needed it.

**Comment by emlun** [comment]
Ready for re-review :)

File: ykman-gui/py/yubikey.py [hunk 9]
```diff
@@ -14,17 +14,18 @@
...
 
 from ykman.descriptor import get_descriptors
+from ykman.driver_ccid import (APDUError, SW)
 from ykman.driver_otp import YkpersError
 from ykman.device import device_config
 from ykman.otp import OtpController
 from ykman.fido import Fido2Controller
-from ykman.piv import PivController
+from ykman.piv import (PivController)
+from ykman.piv import (AuthenticationBlocked, WrongPin, WrongPuk)
```
Comment (dagheyman) [inline ykman-gui/py/yubikey.py]: Drop parenthesis if not multi line import?

File: ykman-gui/qml/YubiKey.qml [hunk 10]
```diff
@@ -306,7 +342,36 @@ Python {
...
+        do_call('yubikey.controller.piv_change_puk', [old_puk, new_puk],
+                _refreshPivBefore(function (resp) {
+                    if (!resp.success && resp.error === 'blocked') {
+                        pivPukBlocked = true
+                    }
+                    cb(resp)
+                }))
+    }
+
     function piv_reset(cb) {
```
Comment (dagheyman) [inline ykman-gui/qml/YubiKey.qml]: Why should reset refresh PIV before?

Comment (emlun) [reply ykman-gui/qml/YubiKey.qml]: The refresh is done before running the callback, after the Python call finishes.

File: ykman-gui/py/yubikey.py (hunk 9 above)
Comment (emlun) [reply ykman-gui/py/yubikey.py]: Sure!

**Comment by emlun** [comment]
Oops! Good catch, should be easy to fix.

File: ykman-gui/py/yubikey.py (hunk 6 above)
Comment (emlun) [reply ykman-gui/py/yubikey.py]: Done!

File: ykman-gui/qml/YubiKey.qml (hunk 7 above)
Comment (emlun) [reply ykman-gui/qml/YubiKey.qml]: Agreed offline to hold off on this for now.

**Comment by emlun** [comment]
Should be fixed now.

**Comment by dagheyman** [comment]
Confirmed that it's fixed.

One more (minor) thing:

When I change FIDO2 PIN and use the wrong PIN, the console prints:
```
qrc:///qml/ChangePinView.qml:66: TypeError: Property 'onClearCurrentPinInput' of object [object Object] is not a function
```

**Comment by emlun** [comment]
Thanks, fixed that one too!

**Comment by emlun** [comment]
See also #74

File: ykman-gui/qml/Fido2ChangePinView.qml [hunk 11]
```diff
@@ -26,7 +26,7 @@ ChangePinView {
                 if (resp.error === 'too long') {
                     fido2TooLongError.open()
                 } else if (resp.error === 'wrong pin') {
-                    clearPinInputs()
+                    clearCurrentPinInput()
```
Comment (dagheyman) [inline ykman-gui/qml/Fido2ChangePinView.qml]: Testing this out, I think all PIN inputs should be cleared. Doing it like this allows the user to block the pin by hitting the Change PIN button repeatedly. Why the change?

Comment (emlun) [reply ykman-gui/qml/Fido2ChangePinView.qml]: Hm, ok. It seemed reasonable to me to only reset the field that's wrong. I got annoyed that I had to re-enter all three values when I got the wrong current PIN, but I guess I've been doing that far more than you'd expect anyone would in normal use.

Would an alternative solution to the blocking the PIN problem be to keep the button disabled while the current PIN field is empty?

Comment (dagheyman) [reply ykman-gui/qml/Fido2ChangePinView.qml]: > Would an alternative solution to the blocking the PIN problem be to keep the button disabled while the current PIN field is empty?

Yes it might, at least it makes sense to do anyway. Current PIN should have a minimum of 4 characters to enable the button.

Comment (dagheyman) [reply ykman-gui/qml/Fido2ChangePinView.qml]: Let's try that solution and see how it feels.

Comment (emlun) [reply ykman-gui/qml/Fido2ChangePinView.qml]: Done.

File: ykman-gui/qml/ChangePinView.qml [hunk 12]
```diff
@@ -29,7 +29,8 @@ ColumnLayout {
     readonly property bool pinMatches: newPin.text === confirmPin.text
 
     function validPin() {
-        return (pinMatches) && (chosenPin.length >= minLength)
+        return (currentPin.length >= minLength)
```
Comment (dagheyman) [inline ykman-gui/qml/ChangePinView.qml]: This blocks setting a FIDO2 PIN, since this will always be false in this view I believe.

Comment (emlun) [reply ykman-gui/qml/ChangePinView.qml]: Oops, should be fixed now. (I don't have a FIDO2 device for testing with me right now...)
